streams/src/main/java/org/apache/kafka/streams/processor/internals/TopologyMetadata.java

```diff
@@ -447,11 +457,106 @@ public String getStoreForChangelogTopic(final String topicName) {
+                    if (!allInputTopics.contains(topic)) {
+                        newTopicsByTopology
+                            .computeIfAbsent(topology, t -> new HashSet<>())
+                            .add(topic);
+                    }
+                }
+            }
+
+            if (!duplicateInputTopics.isEmpty()) {
+                throw new IllegalStateException("The following topics are subscribed to " +
+                    "by multiple topologies: " + duplicateInputTopics);
+            } else if (!subscriptionTopicsWithUnknownSource.isEmpty()) {
+                throw new IllegalStateException("The following topics appear in the subscription " +
+                    "but can't be located in the topology: " + subscriptionTopicsWithUnknownSource);
+            }
+
+            updateAndVerifyNewInputTopics(newTopicsByTopology);
+            for (final Map.Entry<String, Set<String>> topology : newTopicsByTopology.entrySet()) {
+                lookupBuilderForNamedTopology(topology.getKey())
+                    .addSubscribedTopicsFromMetadata(topology.getValue(), logPrefix);
+            }
+        } else {
+            if (!topics.equals(sourceTopicCollection())) {
+                log.error("{}Consumer's subscription does not match the known source topics.\n" +
+                        "consumer subscription: {}\n" +
+                        "topics in topology metadata: {}",
+                    logPrefix, topics, sourceTopicCollection());
+                throw new IllegalStateException("Consumer subscribed topics and topology's known topics do not match.");
+            }
+        }
     }
 
-    void addSubscribedTopicsFromAssignment(final List<TopicPartition> partitions, final String logPrefix) {
-        applyToEachBuilder(b -> b.addSubscribedTopicsFromAssignment(partitions, logPrefix));
+    void addSubscribedTopicsFromAssignment(final Map<TaskId, Set<TopicPartition>> tasks, final String logPrefix) {
+        final Map<String, Set<String>> assignedTopicsByTopology = new HashMap<>();
+
+        for (final Map.Entry<TaskId, Set<TopicPartition>> task : tasks.entrySet()) {
+            final String topologyName = getTopologyNameOrElseUnnamed(task.getKey().topologyName());
+            // Skip updating subscription with topics if their topology is not yet known to this client,
+            // the subscription will be updated when the topology is added
+            if (builders.containsKey(topologyName)) {
+                assignedTopicsByTopology
+                    .computeIfAbsent(topologyName, t -> new HashSet<>())
+                    .addAll(task.getValue().stream().map(TopicPartition::topic).collect(Collectors.toSet()));
+            }
+        }
+        for (final Map.Entry<String, Set<String>> assignedTopics : assignedTopicsByTopology.entrySet()) {
+            final Set<String> newTopics =
+                lookupBuilderForNamedTopology(assignedTopics.getKey())
```

Comment:
Ditto here, should we check that before any new topics are added to the topology, it's topic set should be empty?